task/dump-agent/src/main.rs

```diff
@@ -20,83 +20,48 @@ use userlib::*;
 const_assert!(DUMP_READ_SIZE & (DUMP_READ_SIZE - 1) == 0);
 const_assert!(DUMP_READ_SIZE <= 1024);
 
-struct ServerImpl {
-    areas: [DumpArea; 3],
-}
+struct ServerImpl;
 
 #[cfg(not(feature = "no-rot"))]
 task_slot!(SPROT, sprot);
 
+task_slot!(JEFE, jefe);
+
 impl ServerImpl {
-    fn initialize(&self) {
-        let mut next = 0;
-
-        for area in self.areas.iter().rev() {
-            unsafe {
-                let header = area.address as *mut DumpAreaHeader;
-
-                //
-                // We initialize our dump header with deliberately bad magic
-                // to prevent any dumps until we have everything initialized
-                //
-                (*header) = DumpAreaHeader {
-                    magic: DUMP_UNINITIALIZED,
-                    address: area.address,
-                    nsegments: 0,
-                    written: size_of::<DumpAreaHeader>() as u32,
-                    length: area.length,
-                    agent_version: DUMP_AGENT_VERSION,
-                    dumper_version: DUMPER_NONE,
-                    next,
-                }
-            }
+    fn initialize(&self) -> Result<(), DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.initialize_dump_areas()
+    }
 
-            next = area.address;
-        }
+    fn dump_area(&self, index: u8) -> Result<DumpArea, DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.get_dump_area(index)
+    }
 
-        for area in &self.areas {
-            unsafe {
-                let header = area.address as *mut DumpAreaHeader;
-                (*header).magic = DUMP_MAGIC;
-            }
-        }
+    fn claim_dump_area(&self) -> Result<DumpArea, DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.claim_dump_area()
     }
 
     fn add_dump_segment(
         &mut self,
         addr: u32,
         length: u32,
     ) -> Result<(), DumpAgentError> {
-        let area = self.areas[0];
-
-        unsafe {
-            let header = area.address as *mut DumpAreaHeader;
-
-            if (*header).magic != DUMP_MAGIC {
-                panic!("bad dump magic!");
-            }
-
-            let nsegments = (*header).nsegments;
+        let area = self.dump_area(0)?;
```

Comment:
This does an IPC call to `jefe`.

At the end of the IPC call, can the following sequence occur?

- `jefe` decides to return control to a different (higher-priority) task
-  That task crashes (returning to `jefe`)
- `jefe` saves a dump of the crashed task
- `area` is invalidated in this function

(I'm still reading, so this may not be possible!)

Reply:
This wasn't possible, but has been obviated anyway by moving the initialization code into Jefe.